2_observations/src/munge_diversion_data.R

```diff
@@ -0,0 +1,239 @@
+  # convert units
+  # clean up column and row ordering
+  all_dat <- daily %>%
+    mutate(data_type = 'daily',
+           diversion_cms = diversion_mgd*mgd_to_cms) %>%
+    bind_rows(missing_dailies) %>%
+    mutate(res_level_m = res_level_ft*ft_to_m,
+           gross_storage_m3 = gross_storage_mg*mg_to_cm,
+           storage_change_cms = storage_change_mgd*mgd_to_cms,
+           conservation_release_cms = conservation_release_mgd*mgd_to_cms,
+           directed_release_cms = directed_release_mgd*mgd_to_cms,
+           total_release_cms = total_release_mgd*mgd_to_cms,
+           spillway_cms = spillway_mgd*mgd_to_cms,
+           total_runoff_cms = total_runoff_mgd*mgd_to_cms,
+           storage_above_sill_m3 = stor_above_sill_mg*mg_to_cm) %>%
+    arrange(reservoir, date) %>%
+    select(-year, -month, -contains('_mg'), -res_level_ft) %>%
+    select(reservoir, date, diversion_cms, everything()) %>%
+    relocate(source_file, .after = last_col())
+
+  write_csv(all_dat, file = as_data_file(out_ind))
+  gd_put(out_ind)
+}
+
+#########################################
+# Functions used to map over daily files
+#########################################
+
+# map over sheets within a single monthly file
+# pulling out only Pepacton and Cannsonville, but more included in spreadsheeds
+# however, there were some inconsistencies in other Reservoirs, so wanted to
+# limit the edge cases I needed to handle now
+read_and_munge <- function(location, in_file) {
+  full_path <- file.path(location, in_file)
+
+  # print message to give indication of what file we're on
+  message(paste('Importing data from', in_file))
+  my_colnames <- c('dayofmonth', 'precip_in', 'storage_perc',
+                   'res_level_ft', 'gross_storage_mg',
+                   'storage_change_mgd', 'conservation_release_mgd',
+                   'directed_release_mgd', 'total_release_mgd',
+                   'spillway_mgd', 'diversion_mgd', 'total_runoff_mgd')
+  dat <- full_path %>%
+    excel_sheets() %>%
+    .[grepl('cannon|pepac', ., ignore.case = TRUE)] %>%
+    set_names() %>%
+    purrr::map_df(~readxl::read_excel(full_path, sheet = .x,
+                                      range = 'A5:L43',
+                                      col_names = my_colnames), .id = 'reservoir') %>%
+    mutate(filename = in_file)
+
+  return(dat)
+
+}
+
+# go through each water year to read in and bind
+# each monthly file
+map_over_wy_files <- function(sub_dirs) {
+
+  # print message to give indication of what folder we're on
+  message(paste('Importing data from', sub_dirs))
+  # create file path in new sub director
+  wy_path <- file.path(root_dir, sub_dirs)
+  wy_files <- list.files(wy_path)
+  wy_files <- wy_files[!grepl('csv', wy_files, ignore.case = TRUE)]
+
+  if(!length(wy_files) == 12) {
+    warning(paste(sub_dirs, ' does not contain 12 monthly files'))
+  }
+
+  dat2 <- wy_files %>%
+    map_df(~read_and_munge(location = wy_path, in_file = .x)) %>%
```

Comment:
Nice to see all this `purrr`ing - I've only recently come to really appreciate that package, but it's great

Reply:
This is the first time I've used it from scratch -- once I was writing a loop within a loop within a loop....it felt like the right time to make the jump. Still not totally intuitive to me, but love the resulting simplicity of the code!